### ISSUE TITLE
[RFC] Support easy-to-use Storage in Multi-process on the single server.

### ISSUE THREAD

**Issue by keisuke-umezawa** [optuna/optuna#2950]
## Motivation
In the current implementation, there is no option to run Optuna in multi process without setup such as RDB. There is a need for a function that can be executed in a multi process without any special setup **on the single server**.

### Note
For the case of Multi-process on the multiple severs, some users want to run distributed optimization without RDB setting. That issue will be addressed in another issue if the demand is high.

## Description
We will select a solution from the follows:

1. We will not do anything for it
1. Check if the problem in SQLite ( https://github.com/optuna/optuna/issues/820 ) can be solved, and implement and provide it if possible.
1. Implemented and provided within the Optuna module, such as https://github.com/optuna/optuna/pull/2759
1. Implemented and provided using 3rd party module such as Dask like https://github.com/optuna/optuna/pull/2023

The solution will be selected from the following viewpoints, and the scope of change (e.g. whether a major release is required or not) will also be investigated.

- Usability
    - Demand for the functionality
    - Quality of the functionality
    - Environment / conditions that can provide the functionality
- Implementation / maintenance cost

## TODOs
- [x] Select a solution for multiprocess in-memory storage
- [x] Implement it
    - [x] At first, we try to resolve this issue by `2: Check if the problem in SQLite can be solved, and implement and provide it if possible.`
    - [ ] If not, we will try to resolve this issue by others.

**Comment by MaksimEkin** [comment]
Any updates on this one? I am also getting the ```sqlite3.OperationalError: database is locked ``` error when running the study distributed.

**Comment by MaksimEkin** [comment]
Workaround for me was immediately returning the first trial, which removed the error ```sqlite3.OperationalError: database is locked``` from the remaining trials that are running in parallel:
```
def objective(trial):
    if str(vars(trial)['_trial_id']) == '0':
        return 0
```

I am not exactly sure why this fixed it tho.

**Comment by recycletechno** [comment]
Maybe update this nice class (class MultiProcessStorage) to work with current version of optuna? I read that it was good but now it is outdated
https://gist.github.com/sile/2c329d0cb38f35e25a8c7bab44edbb63

**Comment by keisuke-umezawa** [comment]
@recycletechno 
> Check if the problem in SQLite ( sqlite3.OperationalError: database is locked #820 ) can be solved, and implement and provide it if possible. 

At first, we will it, and if it is solved by sqlite, we will provide it for `easy-to-use Storage in Multi-process`.

**Comment by keisuke-umezawa** [comment]
@MaksimEkin 
> Any updates on this one? I am also getting the sqlite3.OperationalError: database is locked error when running multiple trials in a study parallel.

Could you share your environment and situation for the above error? If we use multi-process on the single server with this method https://github.com/optuna/optuna/issues/820#issuecomment-980841426, we rarely meet the error. But, if you still meet the error, we would like to know your situation. Thank you!

**Comment by HideakiImamura** [comment]
@keisuke-umezawa Is this issue completed since #820 has been closed?

**Comment by keisuke-umezawa** [comment]
Yes!

> For the case of Multi-process on the multiple severs, some users want to run distributed optimization without RDB setting. That issue will be addressed in another issue if the demand is high.

 If we want to do something for it, we will create another issue.
